File: app/src/main/java/com/exzell/mangaplayground/utils/DateUtils.java
package com.exzell.mangaplayground.utils;

import java.text.SimpleDateFormat;
import java.time.Duration;
import java.time.format.DateTimeFormatter;
import java.util.Date;

public class DateUtils {

    public static long translateTime(String weirdTime){

        StringBuilder timeBuild = new StringBuilder(weirdTime);
        long time = (System.currentTimeMillis());

        while (!timeBuild.toString().isEmpty() && !timeBuild.toString().trim().equals("ago")){
            time = parseDate(timeBuild, time);
        }


        String date = SimpleDateFormat.getDateInstance(SimpleDateFormat.SHORT).format(new Date(time));

        return time;
    }


    private static long parseDate(StringBuilder time, long oldTime){

        int endIndex = 0;
        int startIndex;
        long multiplierToSecond = 1;

        if(time.toString().contains("second")){
            endIndex = time.indexOf("second");

        }else if(time.toString().contains("minute")){
            endIndex = time.indexOf("minute");
            multiplierToSecond = 60;

        }else if(time.toString().contains("hour")){
            endIndex = time.indexOf("hour");
            multiplierToSecond = 3600;

        }else if(time.toString().contains("day")){
            endIndex = time.indexOf("day");
            multiplierToSecond = 3600 * 24;

        }else if(time.toString().contains("week")){
            endIndex = time.indexOf("week");
            multiplierToSecond = 7 * 3600 * 24;

        }else if(time.toString().contains("month")){
            endIndex = time.indexOf("month");
            multiplierToSecond = 30 * 3600 * 24;

        }else if(time.toString().contains("year")){
            endIndex = time.indexOf("year");
            multiplierToSecond = 365 * 24 * 3600;
        }



        startIndex = time.lastIndexOf(" ", endIndex-2);
        startIndex = startIndex == -1 ? 0 : startIndex;
        String timeStr = time.substring(startIndex, endIndex).trim();
        int timeInSec = Character.isDigit(timeStr.charAt(0)) ? Integer.parseInt(timeStr) : 1;
        oldTime = oldTime - ((timeInSec * multiplierToSecond) * 1000);


        int nextSpaceIndex = time.indexOf(" ", endIndex);
        if(nextSpaceIndex != -1) time.delete(startIndex, nextSpaceIndex+1);
        else time.delete(startIndex, time.length());

        return oldTime;
    }
}

File: app/src/main/java/com/exzell/mangaplayground/utils/DateUtils.kt
package com.exzell.mangaplayground.utils

import java.util.*

/** Turns the calendar back to 00:00 */
fun Calendar.reset(millis: Long? = null): Calendar {
    millis?.let { setTimeInMillis(it) }
    set(Calendar.MINUTE, 0)
    set(Calendar.HOUR_OF_DAY, 0)
    set(Calendar.SECOND, 0)
    set(Calendar.MILLISECOND, 0)
    return this
}

fun String.translateTime(): Long {
    val timeBuild = StringBuilder(this)
    var time = System.currentTimeMillis()
    while (!timeBuild.toString().isEmpty() && timeBuild.toString().trim { it <= ' ' } != "ago") {
        time = parseDate(timeBuild, time)
    }
    return time
}

private fun parseDate(time: StringBuilder, oldTime: Long): Long {
    var oldTimeMod = oldTime
    var endIndex = 0
    var startIndex: Int
    var multiplierToSecond: Long = 1
    if (time.toString().contains("second")) {
        endIndex = time.indexOf("second")
    } else if (time.toString().contains("minute")) {
        endIndex = time.indexOf("minute")
        multiplierToSecond = 60
    } else if (time.toString().contains("hour")) {
        endIndex = time.indexOf("hour")
        multiplierToSecond = 3600
    } else if (time.toString().contains("day")) {
        endIndex = time.indexOf("day")
        multiplierToSecond = (3600 * 24).toLong()
    } else if (time.toString().contains("week")) {
        endIndex = time.indexOf("week")
        multiplierToSecond = (7 * 3600 * 24).toLong()
    } else if (time.toString().contains("month")) {
        endIndex = time.indexOf("month")
        multiplierToSecond = (30 * 3600 * 24).toLong()
    } else if (time.toString().contains("year")) {
        endIndex = time.indexOf("year")
        multiplierToSecond = (365 * 24 * 3600).toLong()
    }
    startIndex = time.lastIndexOf(" ", endIndex - 2)
    startIndex = if (startIndex == -1) 0 else startIndex
    val timeStr = time.substring(startIndex, endIndex).trim { it <= ' ' }
    val timeInSec = if (Character.isDigit(timeStr[0])) timeStr.toInt() else 1
    oldTimeMod -= timeInSec * multiplierToSecond * 1000
    val nextSpaceIndex = time.indexOf(" ", endIndex)
    if (nextSpaceIndex != -1) time.delete(startIndex, nextSpaceIndex + 1) else time.delete(startIndex, time.length)
    return oldTimeMod
}


/**
 * Returns the hours and minutes as string eg 20:00, 04:24
 */
fun Long.getTimeOnly(): String {
    val stampReset = Calendar.getInstance().reset(this).timeInMillis
    val seconds = (this - stampReset) / 1000
    val hours = seconds.floorDiv(3600)
    val minutes = seconds.mod(3600) / 60

    return StringBuilder().apply {
        if (hours < 10) append("0")

        append(hours)

        append(":")

        if (minutes < 10) append("0")

        append(minutes)

    }.toString()
}